### PR TITLE
Support Kubernetes Job to build Docker image

### PR DESCRIPTION
**Pull request by hdt91**
This PR add the basic support to create a job on a K8s cluster running a predefined image. This image contains a script to clone the specified repository, commit, tag, etc. passed as environment variables, then optionally push the newly built image and reports the result log.

/cc @jonmoter @grosser 

### Tasks
- Fix CodeClimate and code style violations
- Create the build job image and push to registry
- Test the whole flow on the staging cluster
- Fix any issue after code review ?

### References
 - Jira link: https://zendesk.atlassian.net/browse/PAAS-141

### Risks
- Level: Low

### CODE REVIEW

File: app/models/docker_builder_service.rb [hunk 1]
```diff
@@ -21,9 +22,35 @@ def run!(image_name: nil, push: false, tag_as_latest: false)
...
+            if (match = line[DOCKER_REPO_REGEX, 1])
+              build.docker_repo_digest = match
+            end
+          end
+        end
+        if build.docker_repo_digest.blank?
+          output.puts "### Failed to get the image digest"
+        end
+
+        build.save!
```
Comment (jonmoter) [inline app/models/docker_builder_service.rb]: Refactor this logic into a separate method, since `run!` is getting pretty big.

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# Execute a job that can build/push Docker images
+# and write progress to the local job output.
+module Kubernetes
+  class JobExecutor
```
Comment (jonmoter) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: Call this something like `BuildJobExecutor`, since it's specifically for executing a job that builds Docker images, rather something that executes jobs generically.

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    attr_reader :job
+    delegate :id, to: :job
+
+    def initialize(output, job:)
+      @output = output
+      @job = job
+    end
+
+    def build!(reference, repo:, tag:, push: false, registry:)
+      job_log = ""
```
Comment (jonmoter) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: Seems strange to have this write to `@output`, and also to return a string with the job output.  Why not just write all output to `@output`?

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        return false, job_log
+      end
+
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::JobConfigFile.new(contents, build_job_config_path).job
+      k8s_job.metadata.name = "docker-build-job-#{SecureRandom.hex(16)}"
+      k8s_job.first_container.args = [@job.project.repository_url, reference, repo, tag, push ? "yes" : "no"]
+      k8s_job.first_container.env = [
+        {name: 'DOCKER_REGISTRY_USER', value: registry[:username] || ""},
+        {name: 'DOCKER_REGISTRY_PASS', value: registry[:password] || ""},
```
Comment (jonmoter) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: Not a good practice to put sensitive information into the ENV of an object. Should really use a secret, or put something into Vault.

Also, Kubernetes has its own mechanism to authenticate Docker with remote registries. We can probably leverage that. http://kubernetes.io/docs/user-guide/images

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      k8s_job = nil
+
+      if !valid_registry_info?(registry)
+        @output.puts "### Registry server should not be empty. Aborting..."
+        return false, job_log
+      end
+
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::JobConfigFile.new(contents, build_job_config_path).job
+      k8s_job.metadata.name = "docker-build-job-#{SecureRandom.hex(16)}"
```
Comment (jonmoter) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: Might be nice to name the job based on the project and build id.  Like "docker-build-help-center-12421", so that it's more obvious what the job is for.

Also, add some labels like the name of the project, and something labeling it as a docker build job, so we can easily query on these kind of jobs.

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      k8s_job.first_container.env = [
+        {name: 'DOCKER_REGISTRY_USER', value: registry[:username] || ""},
+        {name: 'DOCKER_REGISTRY_PASS', value: registry[:password] || ""},
+        {name: 'DOCKER_REGISTRY_EMAIL', value: registry[:email] || ""},
+        {name: 'DOCKER_REGISTRY', value: registry[:serveraddress]}
+      ]
+
+      @output.puts "### Running a remote Docker build job: #{k8s_job.metadata.name}"
+      extension_client.create_job(k8s_job)
+      success, job_obj = wait_for_job_status(k8s_job)
```
Comment (jonmoter) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: Enhancement: use `watch_pod_log` and write the output to the `@output` buffer, so that the end-user gets feedback before the job completes.

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb (hunk 3 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: The intention is to return the specific job log to the caller so it can parse only the necessary information.

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb (hunk 4 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: I can remove this since it is optional. The authentication happens at the job container so just need to change the image. Is there any documentation about the recommended way that we use auth/secret in the cluster, such as just assume the cluster already has the secret object?

File: plugins/kubernetes/app/models/kubernetes/job_executor.rb (hunk 5 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/job_executor.rb]: Makes sense and a nice addition. Is there any guideline about the convention to attach labels to different Kubernetes resources?

**Comment by hdt91** [comment]
There is another pending PR that actually is sort of a superset for this one. What is your opinion on this?
https://github.com/zendesk/samson/pull/1061

**Comment by jonmoter** [comment]
I think that the other PR has some code that is more general-purpose for using Kubernetes jobs.  However, there's also a lot of copy-pasted code there that needs cleanup.

So my hunch is that we'd be best off cleaning up and merging the other PR, and then adapting your code to fit in with that.

Btw, I'm sorry it's taken so long to review and merge your code here.  I've been really heads-down trying to get the cluster running in our staging environments, wrangling chef cookbooks, etc.  So without a stable Kubernetes cluster, it's not useful to merge in code that relies on the cluster for building. (In general, this project has required a lot more DevOps work than Dev work, for better or worse.)

**Comment by hdt91** [comment]
No worries, that PR actually covers a few TODO that I wanted originally but decided not to avoid spilling the story scope and explore where a job might fit in. I would prefer to have that one get merged nicely, since the adoption should be quite smooth based on what I have read. When everything is good, I will clean up this PR and ping you guys again :D

**Comment by grosser** [comment]
FYI job support is ready in #1092 will hopefully be merged shorty, not sure how helpful that will be though ...

**Comment by hdt91** [comment]
I have cleaned up the PR based on your recent commits and #1092. Please take a look. Btw how to view the Code Climate result?

File: plugins/kubernetes/app/models/kubernetes/config_file.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+module Kubernetes
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/config_file.rb]: this should not be necessary ...

File: plugins/kubernetes/app/models/kubernetes/config_verifier.rb [hunk 8]
```diff
@@ -0,0 +1,112 @@
+module Kubernetes
+  class ConfigVerifier
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/config_verifier.rb]: already have this too ...

**Comment by grosser** [comment]
why do we need multiple verifiers 1 should be able to handle all resource types ...

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+      registry[:serveraddress].present?
+    end
+
+    def build_job_config_path
+      ENV['KUBE_BUILD_JOB_FILE'] || File.join(Rails.root, 'plugins', 'kubernetes', 'config', 'build_job.yml')
+    end
+
+    def fill_job_details(k8s_job, build, project, tag:, push: false, tag_as_latest: false)
+      # Fill in some information to easily query the job resource
+      project_name = project.name.downcase.tr('_', '-')
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: use `.permalink` ?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      labels = { project: project_name, role: "docker-build-job" }
+      k8s_job[:metadata][:name] = "docker-build-#{project_name}-#{SecureRandom.hex(7)}"
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: put the build id in there too ?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+      k8s_job[:metadata][:labels].update(labels)
+      k8s_job[:spec][:template][:metadata][:labels].update(labels)
+
+      # Necessary to query the pod running the job later
+      k8s_job[:spec][:template][:metadata][:labels][:jobName] = k8s_job[:metadata][:name]
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: I think so far we used underscopre labels -> job_name

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+      # Pass all necessary information so that remote container can build the image
+      container_params = {
+        env: [{name: 'DOCKER_REGISTRY', value: @registry[:serveraddress] }],
+        args: [
+          project.repository_url, build.git_sha, project.docker_repo,
+          tag, push ? "yes" : "no", tag_as_latest ? "yes" : "no"
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: this looks a bit crazy ... maybe parentheses can help  tag, (push ? ...), (tag_as_latest ? ...)

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+          selector = "jobName=#{job.name}" # A build job only has one pod
+          pod = Kubernetes::Api::Pod.new(
+            client.get_pods(namespace: k8s_job[:metadata][:namespace], label_selector: selector).first
+          )
+          # If the pod name for the job is unchanged from previous loops, it is likely that the pod finishes
+          # but the job status is not yet updated. In this case, we will get old/duplicated logs if we start
+          # a new watch stream.
+          if pod_name != pod.name
+            pod_name = pod.name
+            client.watch_pod_log(pod.name, pod.namespace).each do |line|
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: I like the streaming part, should give instant feedback

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+      extension_client.create_job(k8s_job.deep_symbolize_keys)
+      start = Time.now
+      logs = ""
+      pod_name = ""
+
+      loop do
+        sleep TICK # Sleep a bit to wait for the server
+        begin
+          job = Kubernetes::Api::Job.new(
+            extension_client.get_job(k8s_job[:metadata][:name], k8s_job[:metadata][:namespace])
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: maybe extract `job_name` it's used 3 times

File: plugins/kubernetes/config/build_job.yml [hunk 15]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (grosser) [inline plugins/kubernetes/config/build_job.yml]: could add a top-level comment here explaining what it does

**Comment by grosser** [comment]
that looks reasonable ...

things I'd like to change:
 - extract as much view logic as possible into new plugin hooks
 - ideally reuse existing verifier/config loader
 - ideally extract remove_job logic into kubernetes ... but that might be tricky ...
 - maybe rename remote_job attribute -> kubernetes_job for clarity

**Comment by hdt91** [comment]
Thanks for all those suggestions. May I know more about what is your thought on the third point?

**Comment by grosser** [comment]
kubernetes related things should live in the kubernetes plugin 
 - for simpler grepping/overview
 - to not distract people working on 'regular' samson
 - to be able to remove it if kubernetes ever goes out of fashion
 - to be able to use the same hooks to plug in another container framework

**Comment by hdt91** [comment]
I have reused the config reader/verifier, cleaned up the variable name and moved the view parts into kubernetes plugin directory.

It is indeed a bit tricky to move the part where the build job execution is done in the docker image service as there is a coupling in parsing the output data and we are not having the full picture of a remote job system. For now the new code chunk looks fairly isolated and can be taken out later when this local build service can be moved out as an external plugin and we have a better integration of remote job concepts.

File: app/controllers/builds_controller.rb [hunk 16]
```diff
@@ -92,8 +92,18 @@ def edit_build_params
...
+      hash[p] = params[p].present?
+    end
+    hash
+  end
+
   def start_docker_build
-    DockerBuilderService.new(@build).run!(push: true)
+    docker_build_params = { push: true }
+    docker_build_params.update(extra_docker_build_params)
+    DockerBuilderService.new(@build).run!(docker_build_params)
```
Comment (grosser) [inline app/controllers/builds_controller.rb]: this can be part of the plugin via a `build_params` hook

File: app/controllers/builds_controller.rb [hunk 17]
```diff
@@ -92,8 +92,18 @@ def edit_build_params
     params.require(:build).permit(:label, :description)
   end
 
+  def extra_docker_build_params
+    hash = {}
+    Samson::Hooks.fire(:extra_build_params).each do |p|
+      hash[p] = params[p].present?
```
Comment (grosser) [inline app/controllers/builds_controller.rb]: this sets true to some keys ... looks strange ... could merge extra_build_params ?

File: app/models/docker_builder_service.rb [hunk 18]
```diff
@@ -32,6 +35,31 @@ def run!(image_name: nil, push: false, tag_as_latest: false)
     JobExecution.start_job(job_execution)
   end
 
+  def run_build_image_job(local_job, image_name, push: false, tag_as_latest: false)
```
Comment (grosser) [inline app/models/docker_builder_service.rb]: can this code live somewhere else ? (not mix kubernetes into samson core)

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+        return false, job_log
+      end
+      @registry = registry
+
+      # Read the external config path and create a new job config instance
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::RoleConfigFile.new(contents, build_job_config_path).job
+      fill_job_details(k8s_job, build, project, tag: tag, push: push, tag_as_latest: tag_as_latest)
+
+      if k8s_job
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: if these is no job file this would all blow up ... now if needed ?

also these past 5-6 lines can be a method ?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+      end
+
+      @output.puts "### Running a remote Docker build job: #{job_name}"
+      success, job_log = create_and_wait_for_job(k8s_job, @output)
+
+      if success
+        @output.puts "### Remote build job #{job_name} completed sucessfully"
+      else
+        @output.puts "### Remote build job #{job_name} failed to run or timed out"
+      end
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: little dry-up with `message = (success ? "completed successfully" : "failed or timed out")`

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+    delegate :id, to: :job
+
+    def initialize(output, job:)
+      @output = output
+      @job = job
+      @registry = {}
+    end
+
+    def execute!(build, project, tag:, push: false, registry:, tag_as_latest: false)
+      job_log = job_name = job_namespace = ""
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: will be nil anyway ... no need to assign ?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+      else
+        @output.puts "### Remote build job #{job_name} failed to run or timed out"
+      end
+
+      return success, job_log
+    ensure
+      if k8s_job
+        # Jobs will still be there regardless of their statuses
+        @output.puts "### Cleaning up the remote build job: #{job_name}"
+        extension_client.delete_job(job_name, job_namespace)
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: this wil blow up with a 404 when the job failed to create ?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+    end
+
+    def create_and_wait_for_job(k8s_job, output)
+      extension_client.create_job(k8s_job.deep_symbolize_keys)
+      start = Time.now
+      logs = pod_name = ""
+      job_name = k8s_job[:metadata][:name]
+      job_namespace = k8s_job[:metadata][:namespace]
+
+      loop do
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: does a job have multiple pods ? ... not sure what's going on here ... some docs might help :)

File: app/controllers/builds_controller.rb (hunk 16 above)
Comment (hdt91) [inline app/controllers/builds_controller.rb]: What do you mean by ```build_params``` hook?

File: app/controllers/builds_controller.rb (hunk 17 above)
Comment (hdt91) [inline app/controllers/builds_controller.rb]: This only does the minimum since for this task there is only one extra argument. If there are extra arguments the function can change to suit the need with extra conversions from string values to proper parameter values.

File: app/models/docker_builder_service.rb (hunk 18 above)
Comment (hdt91) [inline app/models/docker_builder_service.rb]: I would prefer to do this as well but as I said in the earlier comment, this piece of code has a coupling with ```DockerBuilderService``` in parsing the final output and to certain extent kinda of intertwine a local running job and a remote running job with logs content and job status passed around.

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 21 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: Just my habit to declare all variables before using them to deal with code coverage or normal static analysis tool. Is this a bad thing in Ruby/Rails?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 19 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: If there is no job file some succeeding code will likely trigger an exception and be caught in the ```ensure``` block. Is there anything wrong that I am not aware of?

What is your thought on grouping other lines of code into a method? These line look like an one-time glue code between the ```fill_job_details``` and ```create_and_wait_for_job``` to me.

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 22 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: Yes, if it fails to create it will just throw 404 and be caught by the outer try/catch block of job executor, sending a proper message to the log. This ```ensure``` block is the best effort attempt to clean up resource remnants from the build job, what else do you think is good to add here?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 23 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: In this context I mean the build job should only have one container for simplicity. I do not know what kind of use case we need a build job with more than one container.

File: app/controllers/builds_controller.rb (hunk 17 above)
Comment (grosser) [inline app/controllers/builds_controller.rb]: atm it only supports booleans which is a bit strange ...

Comment (grosser) [inline app/controllers/builds_controller.rb]: idk ... good enough for now ...

File: app/views/builds/new.html.erb [hunk 24]
```diff
@@ -40,6 +40,8 @@
         </div>
       </div>
 
+      <%= Samson::Hooks.render_views(:build_new_use_kubernetes_job, self, form: form) %>
```
Comment (grosser) [inline app/views/builds/new.html.erb]: hook names should be universal, something like build_new for example

File: app/views/builds/show.html.erb [hunk 25]
```diff
@@ -61,6 +61,7 @@
   <% unless docker_build_running? @build %>
     <%= form_for :build, url: build_docker_image_project_build_path(@build.project, @build), html: { class: "form-inline" } do |form| %>
       <%= form.submit (job ? 'Rebuild Docker Image' : 'Build Docker Image'), class: "btn btn-primary" %>
+      <%= Samson::Hooks.render_views(:build_show_use_kubernetes_job, self, form: form) %>
```
Comment (grosser) [inline app/views/builds/show.html.erb]: same here -> build_show

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 21 above)
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: nah not really just looks a bit strange ...

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 19 above)
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: 28-35 looks like one concern to me something like 'job_config'

Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: the File.read will already blow up, so I don't think we need a if k8s_job check ...

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 22 above)
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: blowing up in the ensure will clobber the original error message, so I'd prefer a begin; delete_job; rescue KubeClient::Exception;end or so to silence it

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 23 above)
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: not asking about containers, but why we need to loop here ... I was thinking we'd only have 1 pod created / don't need any looping logic

File: plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb [hunk 26]
```diff
@@ -0,0 +1,152 @@
+require_relative "../../test_helper"
+
+SingleCov.covered! uncovered: 3
```
Comment (grosser) [inline plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb]: can you add a comment on why there are uncovered lines or possibly add some test to hit them ?

File: plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+
+        it 'job is unsuccessful' do
+          job_api_obj.stubs(:failure?).returns true
+
+          success, job_log = execute!(registry_info: registry_info)
+          refute success
+          assert_empty job_log
+        end
+
+        it 'job gets timeout' do
```
Comment (grosser) [inline plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb]: strange test case name ... how about something like `it does xyz when the job times out`

File: plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+          assert_equal v[:role], 'docker-build-job'
+          assert_equal v[:foo], 'bar'
+        end
+
+        assert_equal build_config.job[:spec][:template][:spec][:containers][0][:args],
+          [project.repository_url, build.git_sha, project.docker_repo, 'latest', 'no', 'no']
+        assert_equal build_config.job[:spec][:template][:spec][:containers][0][:env].length, 1
+      end
+
+      describe 'can create job' do
```
Comment (grosser) [inline plugins/kubernetes/test/models/kubernetes/build_job_executor_test.rb]: describes usually are something about a certain state like "describe 'when all params are good' do ... it 'creates a job'"

File: plugins/kubernetes/test/test_helper.rb [hunk 29]
```diff
@@ -106,7 +106,11 @@ def kubernetes_fake_raw_template
     Kubernetes::ReleaseDoc.any_instance.stubs(raw_template: read_kubernetes_sample_file('kubernetes_deployment.yml'))
   end
 
+  def sample_file_path(file_name)
```
Comment (grosser) [inline plugins/kubernetes/test/test_helper.rb]: this is a global method so it needs to be namespaced to kubernetes ... kubernetes_sample_file_path

File: test/models/docker_builder_service_test.rb [hunk 30]
```diff
@@ -1,6 +1,6 @@
 require_relative '../test_helper'
 
-SingleCov.covered! uncovered: 19
+SingleCov.covered! uncovered: 20
```
Comment (grosser) [inline test/models/docker_builder_service_test.rb]: what line did not get covered ... can we hit it somehow ?

File: test/models/docker_builder_service_test.rb [hunk 31]
```diff
@@ -23,6 +23,33 @@
...
+    let(:local_job) { stub }
+    let(:k8s_job) { stub }
+    let(:repo_digest) { 'sha256:5f1d7c7381b2e45ca73216d7b06004fdb0908ed7bb8786b62f2cdfa5035fde2c' }
+    let(:build_log) { ["status: Random status", "BUILD DIGEST: #{project.docker_repo}@#{repo_digest}"].join("\n") }
+
+    before do
+      Kubernetes::BuildJobExecutor.expects(:new).returns k8s_job
+    end
+
+    it 'calls the job executor and the job completes successfully' do
```
Comment (grosser) [inline test/models/docker_builder_service_test.rb]: something more descriptive like `it "does xyz when the job completes"`

File: test/models/docker_builder_service_test.rb [hunk 32]
```diff
@@ -23,6 +23,33 @@
...
+      k8s_job.expects(:execute!).returns([true, build_log])
+      build.label = "Version 123"
+
+      service.run_build_image_job(local_job, nil)
+      assert_equal('version-123', build.docker_ref)
+      assert_equal("#{project.docker_repo}@#{repo_digest}", build.docker_repo_digest)
+    end
+
+    it 'calls the job executor but it is not successful' do
```
Comment (grosser) [inline test/models/docker_builder_service_test.rb]: it 'does xyz when the job fails'

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb (hunk 23 above)
Comment (hdt91) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: Got it. I added a comment why it needs a loop. Sometimes the job is not yet ready and the pod log is not available so we need to retry a few seconds later.

File: test/models/docker_builder_service_test.rb (hunk 30 above)
Comment (hdt91) [inline test/models/docker_builder_service_test.rb]: The ```run!``` method's body is uncovered and originally had 19 lines. I added an if check for running the remote job and now the body has one extra line. Do you have any idea how to test the code block passed to the job execution inside this method's body?

Comment (grosser) [inline test/models/docker_builder_service_test.rb]: nah ... just sanity checking whenever this increases ... you are welcome to
test it, but that can be a separate PR :)

**Comment by hdt91** [comment]
I updated this PR with better test case description and other minor cleanups. Still there is a red flag in CodeClimate, do you mind checking what is the issue there?

**Comment by grosser** [comment]
"Analysis results are provided for code written by licensed individuals, and Hai Dang Tran does not appear to be a member of your organization ." ... so no idea ... maybe it's just not doing the report and that's why it's red ...

File: app/controllers/builds_controller.rb (hunk 16 above)
Comment (grosser) [inline app/controllers/builds_controller.rb]: nvm ... build_params would be attributes for the build ... this is something else ...

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+        extension_client.delete_job(job_name, job_namespace) if k8s_job
+      rescue KubeException
+        @output.puts "### Failed to clean up the remote build job"
+      end
+    end
+
+    private
+
+    def build_cluster
+      Kubernetes::Cluster.first!
```
Comment (grosser) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: this most likely needs some tweeking ... something like picking a context / cluster via env vars ... but good enough for now ...

File: plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb [hunk 34]
```diff
@@ -0,0 +1,4 @@
+<div class="form-group">
+  <%= check_box_tag :kubernetes_job%>
+  <%= form.label :kubernetes_job, "Use Remote Kubernetes Job", class: "control-label" %>
```
Comment (grosser) [inline plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb]: should this be a column instead ? ... then we don't need extra_build_params ... just build_params and let it persist ... and no additional UI for build/show ...

**Comment by grosser** [comment]
looking good ❤️

File: plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb (hunk 34 above)
Comment (hdt91) [inline plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb]: If this is persisted as a build's parameter, it becomes a part of core Samson workflow, doesn't it?

Comment (grosser) [inline plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb]: nah ... just an extra column (default false) nobody needs to know about :)

Comment (hdt91) [inline plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb]: Sounds good to me. Only one small issue is to switch between local build and remote build option without an UI. What is your suggestion?

Comment (grosser) [inline plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb]: the new build form gets checkbox ... rebuild does not really need it ... can always make a new build if someone wants to build without remote

Comment (hdt91) [inline plugins/kubernetes/app/views/samson_kubernetes/_build_show.erb]: Updated with the new column :)

**Comment by grosser** [comment]
looks good to me ... @jonmoter / @rata want to give this a sanity check before I merge ?

**Comment by jonmoter** [comment]
Just did a quick review.  Nothing jumped out to me as a blocker.  Let's merge, try it out for a bit, and iterate if need be.

👍

**Comment by hdt91** [comment]
Thanks for very quick feedback guys :D

File: app/models/docker_builder_service.rb [hunk 35]
```diff
@@ -91,6 +119,20 @@ def project
     @build.project
   end
 
+  def registry_credentials
+    return nil unless ENV['DOCKER_REGISTRY'].present?
+    {
+      username: ENV['DOCKER_REGISTRY_USER'],
+      password: ENV['DOCKER_REGISTRY_PASS'],
+      email: ENV['DOCKER_REGISTRY_EMAIL'],
```
Comment (rata) [inline app/models/docker_builder_service.rb]: email is deprecated and will be removed in docker 1.13 IIRC. Maybe we can remove it now?

File: plugins/kubernetes/app/models/kubernetes/build_job_executor.rb [hunk 36]
```diff
@@ -0,0 +1,144 @@
...
+
+    def build_job_config_path
+      ENV['KUBE_BUILD_JOB_FILE'] || File.join(Rails.root, 'plugins', 'kubernetes', 'config', 'build_job.yml')
+    end
+
+    def fill_job_details(k8s_job, build, project, tag:, push: false, tag_as_latest: false)
+      # Fill in some information to easily query the job resource
+      project_name = project.permalink.tr('_', '-')
+      labels = { project: project_name, role: "docker-build-job" }
+      k8s_job[:metadata][:name] = "#{project_name}-docker-build-#{build.id}-#{SecureRandom.hex(7)}"
```
Comment (rata) [inline plugins/kubernetes/app/models/kubernetes/build_job_executor.rb]: why not generateName, instead of name, that does this?

**Comment by rata** [comment]
@grosser thanks, seems fine!
